File: 704. 二分查找/solution.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        left = 0
        right = len( nums) - 1  # 闭区间[left, right]
        if target < nums[left] or target > nums[right]:
            return -1
        while left <= right:  #区间不为空
            # left + (right - left) / 2 在c++和java中可以不溢出
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1 #[mid+1, right]
            else:
                right = mid - 1 #[left, mid-1]
        return left if nums[left] == target else -1
    
nums = [-1,0,3,5,9,12]
target = 9
solution = Solution()
print(solution.search(nums, target))

File: 704. 二分查找/solution.java
class Main {
    public static void main(String[] args) {
        solution solution = new solution();
        
        // 测试用例1: 目标值在数组中
        int[] nums1 = {-1,0,3,5,9,12};
        int target1 = 9;
        System.out.println("index = " + solution.search(nums1, target1));
    }
}

class solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int l = 0, r = n - 1;
        if(target < nums[l] || target > nums[r]){
            return -1;
        }
        while(l <= r){
            int mid = l + (r - l) / 2;
            if(nums[mid] < target){
                l = mid + 1;
            }
            else{
                r = mid - 1;
            }

        }
        if(nums[l] != target){
            return -1;
        }
        else{
            return l;
        }

    }
}